content/intro-to-storybook/vue/zh-CN/composite-component.md

```diff
@@ -0,0 +1,244 @@
++     <div class="wrapper-message">
++       <span class="icon-check" />
++       <div class="title-message">You have no tasks</div>
++       <div class="subtitle-message">Sit back and relax</div>
++     </div>
+    </div>
+
+    <template v-else>
++     <Task v-for="task in tasksInOrder" :key="task.id" :task="task" v-on="$listeners" />
+    </template>
+  </div>
+</template>
+
+<script>
+  import Task from './Task';
+  export default {
+    name: 'TaskList',
+    components: { Task },
+    props: {
+      tasks: { type: Array, required: true, default: () => [] },
+      loading: { type: Boolean, default: false },
+    },
+    computed: {
++     tasksInOrder() {
++       return [
++         ...this.tasks.filter(t => t.state === 'TASK_PINNED'),
++         ...this.tasks.filter(t => t.state !== 'TASK_PINNED'),
++       ];
++     },
+      isEmpty() {
+        return this.tasks.length === 0;
+      },
+    },
+  };
+</script>
+```
+
+通过上述代码我们生成了如下 UI：
+
+<video autoPlay muted playsInline loop>
+  <source
+    src="/intro-to-storybook/finished-tasklist-states.mp4"
+    type="video/mp4"
+  />
+</video>
+
+注意固定项目出现在列表中的位置。我们希望固定项目可以出现在列表的顶端以提示用户其优先度。
+
+## 自动化测试
+
+在上一章中我们学习了如何使用 Storyshots 来进行快照测试。`Task`没有太多的复杂性，所以已经够用了。对于`TaskList`来说其增加了另一层复杂性，这就需要我们寻找一个合适的自动化测试方法来验证特定的输入可以产生特定的输出。我们使用[Jest](https://facebook.github.io/jest/)加测试渲染器来创建单元测试。
+
+![Jest logo](/intro-to-storybook/logo-jest.png)
+
+### 使用 Jest 进行单元测试
+
+Storybook 使用手动检查和快照测试的方式来防止 UI 的 bug。看起来好像只要我们覆盖了足够多的场景，并且使用一些工具保证可以人为检查 story 的变化后，错误将会大大减少。
+
+但是，魔鬼存在于细节中。我们还需要一个测试框架来显示的实现上述需求。也就是单元测试。
+
+在我们的例子中，我们希望`TastList`可以将`tasks`属性中的固定 task 渲染在非固定 task 的**上面**。尽管我们已经有一个 story (`WithPinnedTasks`)来对应此场景；但对于任何人为的检查来说仍然过于含糊了，因此当组件**停止**按上述那样排序时 bug 就产生了。很显然这样的测试并不会大声告诉您**你错了！**。
+
+所以为了防止这样的问题发生，我们可以使用 Jest 将 story 渲染成 DOM，然后跑一些 DOM 查询代码来验证输出中重要的部分。story 的 format 非常棒的一点在于，我们只需要简单的导入 story 到我们的测试中就可以渲染它了！
+
+创建一个测试文件`tests/unit/TaskList.spec.js`。我们创建测试来判断输出结果。
+
+```js:title=tests/unit/TaskList.spec.js
+import Vue from 'vue';
+
+import TaskList from '../../src/components/TaskList.vue';
+
+//👇 Our story imported here
+import { WithPinnedTasks } from '../../src/components/TaskList.stories';
+
+it('renders pinned tasks at the start of the list', () => {
+  // render Tasklist
+  const Constructor = Vue.extend(TaskList);
+  const vm = new Constructor({
+    //👇 Story's args used with our test
+    propsData: WithPinnedTasks.args,
+  }).$mount();
+  const firstTaskPinned = vm.$el.querySelector('.list-item:nth-child(1).TASK_PINNED');
+
+  // We expect the pinned task to be rendered first, not at the end
+  expect(firstTaskPinned).not.toBe(null);
+});
+```
+
+![TaskList test runner](/intro-to-storybook/tasklist-testrunner.png)
+
+注意我们可以在 story 和单元测试中重用`withPinnedTasksData`；这样我们就可以继续以越来越多的方式运用现有资源（代表组件各种有趣配置的示例）。
+
+请注意这个测试仍然十分脆弱。随着项目逐渐成熟，以及`Task`的实现改变时 -- 也许使用了另一个类名 -- 那测试很可能失败，这需要我们去更新它。这不一定是个问题，但是在 UI 中使用单元测试仍需十分小心。它们的维护工作并不容易。除了依靠肉眼，快照和视觉回归（visual regression）（参见[测试章节/intro-to-storybook/vue/zh-CN/test/)）。
```

Comment:
@zydmayday it seems that this needs a bit of work. I was checking the preview associated with your pull request and this link wasn't rendering correctly. It's actually showing the full text.

Reply:
I fixed the link issue.

I am not sure if it is deliberately just use the left parenthesis in the last line of composite-component.md.  
refer here: https://github.com/chromaui/learnstorybook.com/blob/master/content/intro-to-storybook/vue/en/composite-component.md

I added a right parenthesis also.